xml and jsonpath support

bdraco opened the pull request:
https://github.com/home-assistant/home-assistant/pull/31809

## Proposed change
<!-- 
    Describe the big picture of your changes here to communicate to the
    maintainers why we should accept this pull request. If it fixes a bug
    or resolves a feature request, be sure to link to that issue in the 
    additional information section.
-->



## Type of change
<!--
    What types of changes does your PR introduce to our documentation/website?
    Put an `x` in the boxes that apply. You can also fill these out after
    creating the PR.
-->

- [ ] Spelling, grammar or other readability improvements (`current` branch).
- [ ] Adjusted missing or incorrect information in the current documentation (`current` branch).
- [ ] Added documentation for a new integration I'm adding to Home Assistant (`next` branch).
- [x] Added documentation for a new feature I'm adding to Home Assistant (`next` branch).
- [ ] Removed stale or deprecated documentation.

## Additional information
<!--
    Details are important, and help maintainers processing your PR.
    Please be sure to fill out additional details, if applicable.
-->

- Link to parent pull request in the codebase: https://github.com/home-assistant/home-assistant/pull/31809
- This PR fixes or closes issue:

## Checklist
<!--
    Put an `x` in the boxes that apply. You can also fill these out after
    creating the PR. If you're unsure about any of them, don't hesitate to ask.
    We're here to help! This is simply a reminder of what we are going to look
    for before merging your code.
-->

- [x] This PR uses the correct branch, based on one of the following:
  - I made a change to the existing documentation and used the `current` branch.
  - I made a change that is related to an upcoming version of Home Assistant and used the `next` branch.
- [x] The documentation follows the Home Assistant documentation [standards][].

[standards]: https://developers.home-assistant.io/docs/documentation_standards.html

bdraco committed:
xml and jsonpath support

https://github.com/home-assistant/home-assistant/pull/31809

bdraco committed:
adjust for lint

bdraco committed:
grammar

arsaboo commented on source/_integrations/rest.markdown:
```diff
@@ -264,6 +268,24 @@ sensor:
+      - city
+      - zipcode
+    resource: https://jsonplaceholder.typicode.com/users
+    value_template: '{{ value_json.name }}'
```
Should be `{{ value_json[0].name }}`

bdraco replied:
Thanks @arsaboo.  Will adjust this weekend as I'm traveling today.

bdraco committed:
Fix xml conversion output, add a more complex example

bdraco committed:
Update rest.markdown

arsaboo commented:
While you are at it, can you also provide little more details on how to obtain `json_attributes_path`? Can you use the new example that you added to clarify the same?

bdraco committed:
Added note about json_attributes_path

bdraco commented:
> While you are at it, can you also provide little more details on how to obtain `json_attributes_path`? Can you use the new example that you added to clarify the same?

Added.

arsaboo reviewed (dismissed):
Thanks. I am thinking we may need a couple more examples, but that can be done over time.

arsaboo commented:
One final comment...can you also link to the https://xmltodict-debugger.glitch.me/ that folks can use to debug. I added it to the release post, but it will be good to have it here.

frenck commented on source/_integrations/rest.markdown:
```diff
@@ -105,12 +105,16 @@ headers:
+  required: false
   type: [string, list]
+json_path_attributes:
+  description: A [jsonpath](https://goessner.net/articles/JsonPath/) that refereances the location of the json_attributes in the JSON content.
```
```suggestion
  description: A [JSONPath](https://goessner.net/articles/JsonPath/) that references the location of the `json_attributes` in the JSON content.
```

frenck commented on source/_integrations/rest.markdown:
````diff
@@ -264,6 +268,24 @@ sensor:
 ```
 {% endraw %}
 
+[JSONPlaceholder](https://jsonplaceholder.typicode.com/) provides sample JSON data for testing. In the below example, jsonpath locates the attributes in the JSON document. [JSONPath Online Evaluator](https://jsonpath.com/) provides a tool to test your jsonpath. If the endpoint returns XML, it will be converted to JSON using xmltodict before searching for attributes.
````
```suggestion
[JSONPlaceholder](https://jsonplaceholder.typicode.com/) provides sample JSON data for testing. In the below example, JSONPath locates the attributes in the JSON document. [JSONPath Online Evaluator](https://jsonpath.com/) provides a tool to test your JSONPath. If the endpoint returns XML, it will be converted to JSON using `xmltodict` before searching for attributes.
```

frenck commented on source/_integrations/rest.markdown:
```diff
@@ -264,6 +268,24 @@ sensor:
 
+[JSONPlaceholder](https://jsonplaceholder.typicode.com/) provides sample JSON data for testing. In the below example, jsonpath locates the attributes in the JSON document. [JSONPath Online Evaluator](https://jsonpath.com/) provides a tool to test your jsonpath. If the endpoint returns XML, it will be converted to JSON using xmltodict before searching for attributes.
+
+{% raw %}
```
```suggestion
{% raw %}

```

frenck commented on source/_integrations/rest.markdown:
````diff
@@ -264,6 +268,24 @@ sensor:
+    resource: https://jsonplaceholder.typicode.com/users
+    value_template: '{{ value_json[0].name }}'
+```
+{% endraw %}
````
```suggestion

{% endraw %}
```

frenck commented on source/_integrations/rest.markdown:
````diff
@@ -358,3 +380,112 @@ sensor:
 ```
 {% endraw %}
+
+The below example allows shows how to extract multiple values from a dictionary with `json_attributes` and `json_attributes_path` from the XML of a Steamist Steambath WiFi interface and use them to create a switch and multiple sensors without having to poll the endpoint numerous times.
````
```suggestion
The below example allows shows how to extract multiple values from a dictionary with `json_attributes` and `json_attributes_path` from the XML of a Steamist Steambath Wi-Fi interface and use them to create a switch and multiple sensors without having to poll the endpoint numerous times.
```

frenck commented on source/_integrations/rest.markdown:
```diff
@@ -358,3 +380,112 @@ sensor:
+
+The below example allows shows how to extract multiple values from a dictionary with `json_attributes` and `json_attributes_path` from the XML of a Steamist Steambath WiFi interface and use them to create a switch and multiple sensors without having to poll the endpoint numerous times.
+
+In the below example `json_attributes_path` is set to "$.response" which is the location of the the "usr0", "pot0", ... attributes used for `json_attributes`.
```
```suggestion
In the below example `json_attributes_path` is set to `$.response` which is the location of the `usr0`, `pot0`, ... attributes used for `json_attributes`.
```

frenck commented on source/_integrations/rest.markdown:
```diff
@@ -358,3 +380,112 @@ sensor:
+
+In the below example `json_attributes_path` is set to "$.response" which is the location of the the "usr0", "pot0", ... attributes used for `json_attributes`.
+
+{% raw %}
```
```suggestion
{% raw %}

```

frenck commented on source/_integrations/rest.markdown:
```diff
@@ -358,3 +380,112 @@ sensor:
+rest_command:  
+  set_steam_led:
+    url: http://192.168.1.105/leds.cgi?led={{ led }}
+
```
```suggestion
```

frenck commented on source/_integrations/rest.markdown:
````diff
@@ -358,3 +380,112 @@ sensor:
+    url: http://192.168.1.105/leds.cgi?led={{ led }}
+
+```
+{% endraw %}
````
```suggestion

{% endraw %}
```

frenck commented on source/_integrations/rest.markdown:
```diff
@@ -358,3 +380,112 @@ sensor:
+
+For reference, the XML content of endpoint shown above example is below:
+
+{% raw %}
```
```suggestion
```

frenck commented on source/_integrations/rest.markdown:
```diff
@@ -358,3 +380,112 @@ sensor:
+	<usr0>0</usr0>
+	<temp0>0x73XF0x73XF</temp0>
+	<time0> 0</time0>
+
```
we don't need that many new lines here. Let's keep it tidy and make it a single new-line.

bdraco replied:
fixed

frenck commented on source/_integrations/rest.markdown:
````diff
@@ -358,3 +380,112 @@ sensor:
+	
+ </response>
+```
+{% endraw %}
````
```suggestion
```

frenck reviewed (changes requested):
Thank you for the PR @bdraco!

I've reviewed it, and left some comments. Could you please take a look? Thanks! 👍

bdraco committed:
Update source/_integrations/rest.markdown

Co-Authored-By: Franck Nijhof <frenck@frenck.nl>

bdraco committed:
Update source/_integrations/rest.markdown

Co-Authored-By: Franck Nijhof <frenck@frenck.nl>

bdraco committed:
Update source/_integrations/rest.markdown

Co-Authored-By: Franck Nijhof <frenck@frenck.nl>

bdraco committed:
Update source/_integrations/rest.markdown

Co-Authored-By: Franck Nijhof <frenck@frenck.nl>

bdraco committed:
Update source/_integrations/rest.markdown

Co-Authored-By: Franck Nijhof <frenck@frenck.nl>

bdraco committed:
Update source/_integrations/rest.markdown

Co-Authored-By: Franck Nijhof <frenck@frenck.nl>

bdraco committed:
Update source/_integrations/rest.markdown

Co-Authored-By: Franck Nijhof <frenck@frenck.nl>

bdraco committed:
Update source/_integrations/rest.markdown

Co-Authored-By: Franck Nijhof <frenck@frenck.nl>

bdraco committed:
Update source/_integrations/rest.markdown

Co-Authored-By: Franck Nijhof <frenck@frenck.nl>

bdraco committed:
Update rest.markdown

bdraco committed:
Update source/_integrations/rest.markdown

Co-Authored-By: Franck Nijhof <frenck@frenck.nl>

bdraco committed:
Update source/_integrations/rest.markdown

Co-Authored-By: Franck Nijhof <frenck@frenck.nl>

bdraco committed:
Update rest.markdown

bdraco commented:
> One final comment...can you also link to the https://xmltodict-debugger.glitch.me/ that folks can use to debug. I added it to the release post, but it will be good to have it here.

added

arsaboo commented:
@bdraco Can you take a look at this (as another potential example to be included in the future) https://github.com/home-assistant/home-assistant/issues/32064

frenck reviewed (approved):
LGTM ✅ 

@arsaboo Could you do a final check and merge when happy? 👍

bdraco commented:
> @bdraco Can you take a look at this (as another potential example to be included in the future) [home-assistant/home-assistant#32064](https://github.com/home-assistant/home-assistant/issues/32064)

Will do in a followup

dynasticorpheus commented:
@bdraco Any pressing reasons to do json conversion only if an endpoint specifically returns XML with the “text/xml” content type? Could we perhaps relax this a bit and also accept application/xml?

bdraco commented:
@dynasticorpheus https://github.com/home-assistant/home-assistant/pull/32289
